File: MFE.Net.Messaging/NetworkMessage.cs
using System;
using System.Collections;
using System.Ext.Xml;
using System.IO;
using System.Text;
using System.Xml;
using MFE.Utilities;

namespace MFE.Net.Messaging
{
    public class NetworkMessage
    {
        #region Fields
        private string id;
        private Hashtable parameters = new Hashtable();
        #endregion

        #region Properties
        public string ID
        {
            get { return id; }
        }
        public string this[string parameterName]
        {
            get
            {
                return parameters.Contains(parameterName) ? (string)parameters[parameterName] : null;
            }
            set
            {
                if (parameters.Contains(parameterName))
                    parameters[parameterName] = value;
                else
                    parameters.Add(parameterName, value);
            }
        }
        public int ParametersCount
        {
            get { return parameters.Count; }
        }
        #endregion

        #region Constructor
        public NetworkMessage(string id)
        {
            if (Utils.IsStringNullOrEmpty(id))
                throw new ArgumentNullException("NetworkMessage ID not specified");

            this.id = id;
        }
        #endregion

        #region Public Methods
        public static NetworkMessage FromString(string str, NetworkMessageFormat format)
        {
            return (format == NetworkMessageFormat.XML ? FromXml(str) : FromText(str));
        }
        public byte[] Pack(NetworkMessageFormat format)
        {
            return Encoding.UTF8.GetBytes(PackToString(format));
        }
        public string PackToString(NetworkMessageFormat format)
        {
            return NetworkMessageDelimiters.BOM +
                (format == NetworkMessageFormat.XML ? ToXml() : ToText()) +
                NetworkMessageDelimiters.EOM;
        }
        #endregion

        #region Private methods
        private string ToXml()
        {
            MemoryStream ms = new MemoryStream();
            using (XmlWriter writer = XmlWriter.Create(ms))
            {
                //writer.WriteProcessingInstruction("xml", "version=\"1.0\" encoding=\"utf-8\"");

                writer.WriteStartElement("MSG");
                writer.WriteAttributeString("ID", id);

                foreach (string key in parameters.Keys)
                {
                    writer.WriteStartElement("P");
                    writer.WriteAttributeString("Name", key);
                    writer.WriteString((string)parameters[key]);
                    writer.WriteEndElement();
                }

                writer.WriteEndElement();
                writer.Flush();
                writer.Close();
            }

            return new string(Encoding.UTF8.GetChars(ms.ToArray()));
        }
        private string ToText()
        {
            string res = "ID=" + id + ";";
            foreach (string key in parameters.Keys)
                res += key + "=" + (string)parameters[key] + ";";

            return res;
        }

        private static NetworkMessage FromXml(string xml)
        {
            NetworkMessage msg = null;

            byte[] bb = Encoding.UTF8.GetBytes(xml);
            if (bb != null && bb.Length != 0)
            {
                try
                {
                    MemoryStream xmlStream = new MemoryStream(bb);

                    XmlReaderSettings ss = new XmlReaderSettings();
                    ss.IgnoreWhitespace = true;
                    ss.IgnoreComments = true;
                    using (XmlReader reader = XmlReader.Create(xmlStream, ss))
                    {
                        while (!reader.EOF)
                        {
                            reader.Read();
                            switch (reader.NodeType)
                            {
                                case XmlNodeType.Element:
                                    switch (reader.Name)
                                    {
                                        case "MSG":
                                            msg = new NetworkMessage(reader.GetAttribute("ID"));
                                            break;
                                        case "P":
                                            if (msg != null)
                                                msg[reader.GetAttribute("Name")] = reader.ReadString();
                                            break;
                                        default:
                                            break;
                                    }
                                    break;
                                default:
                                    break;
                            }
                        }
                    }
                }
                catch (Exception e)
                {
                    msg = null;
                }
            }

            return msg;
        }
        private static NetworkMessage FromText(string txt)
        {
            NetworkMessage msg = null;

            Hashtable parameters = new Hashtable();

            try
            {
                string[] pairs = txt.Split(new Char[] { ';' });
                foreach (string pair in pairs)
                {
                    if (!Utils.IsStringNullOrEmpty(pair))
                    {
                        string[] s = pair.Split(new Char[] { '=' });
                        parameters.Add(s[0], s[1]);
                    }
                }

                if (parameters.Contains("ID"))
                {
                    msg = new NetworkMessage((string)parameters["ID"]);
                    foreach (string key in parameters.Keys)
                        if (key != "ID")
                            msg[key] = (string)parameters[key];
                }
            }
            catch { }

            return msg;
        }
        #endregion
    }
}

File: Typhoon.HttpFileServerTest/DOTNETMF_FS_EMULATION/WINFS/Script/NetworkMessage.js

function NetworkMessage(ID) {
    var me = this;
    var id = ID;
    var parameters = [];

    this.GetID = function () { return id; }
    this.SetParameter = function (name, value) {
        for (var i = 0; i < parameters.length; i++) {
            var pair = parameters[i];
            if (pair[0] == name) {
                pair[1] = value;
                return;
            }
        }

        parameters.push([name, value]);
    }
    this.GetParameter = function (name) {
        for (var i = 0; i < parameters.length; i++) {
            var pair = parameters[i];
            if (pair[0] == name)
                return pair[1];
        }
        return null;
    }
    this.ToText = function () {
        var res = "";

        res += "<BOM>";
        
        res += "ID=" + id + ";";
        for (var i = 0; i < parameters.length; i++) {
            var pair = parameters[i];
            res += pair[0] + "=" + pair[1] + ";";
        }

        res += "<EOM>";

        return res;
    }
    this.FromText = function (txt) {
        txt = txt.substr(5, txt.length - 10); // remove delimiters

        parameters = [];

        var pairs = txt.split(";");
        if (pairs && pairs.length) {
            for (var i = 0; i < pairs.length; i++) {
                var entry = pairs[i];
                if (entry && entry.length) {
                    var pair = entry.split("=");
                    if (pair && pair.length) {
                        if (pair[0] == "ID")
                            id = pair[1];
                        else
                            this.SetParameter(pair[0], pair[1]);
                    } 
                }
            }
        }
    }
}